File: src/id/handler-signal.ts
import type {
  JSXOpeningElement,
  Expression,
  Node,
} from '@babel/types';
import { slugify } from './slugify';

/**
 * Mine an English intent slug from an element's event handlers.
 *
 * On non-Latin UIs (e.g. Chinese) the visible text carries no ASCII signal, but
 * developers almost always name their handlers, state setters, and call
 * arguments in English (`onClick={() => setDateType("today")}`,
 * `onClick={datePicker.open}`, `onClick={handleSubmit}`). That English *is* the
 * stable, human-meaningful signal — more reliable than the label — so we prefer
 * it over a Unicode fallback.
 *
 * Returns `null` when no handler yields a meaningful slug (so the caller can
 * fall through to the next signal). Position is never consulted here.
 */

/** Event handler props, in priority order, most likely to name the intent. */
const HANDLER_ATTRS = ['onClick', 'onPress', 'onChange', 'onSubmit', 'onValueChange'];

/**
 * Slugs too generic (or too short) to make a useful id. A single-letter handler
 * (`onClick={s}`) or a bare `handle`/`callback` name tells QA nothing, so we
 * reject it and let the caller fall back to the label text.
 */
const STOPWORDS = new Set([
  'fn',
  'cb',
  'callback',
  'handler',
  'handle',
  'on',
  'e',
  'ev',
  'evt',
  'event',
]);

/** Strip the conventional `handle`/`on` prefix: `handleSubmit` -> `Submit`. */
function stripHandlerPrefix(name: string): string {
  const stripped = name.replace(/^(handle|on)(?=[A-Z_])/, '').replace(/^_+/, '');
  return stripped || name;
}

/** Reject empty, single-letter, or stopword slugs. */
function meaningful(slug: string): string | null {
  if (!slug || slug.length < 2) return null;
  if (STOPWORDS.has(slug.replace(/-/g, ''))) return null;
  return slug;
}

/** Narrow to call-like nodes (`foo()` and `foo?.()`). */
function isCall(n: Node): n is Extract<Node, { type: 'CallExpression' | 'OptionalCallExpression' }> {
  return n.type === 'CallExpression' || n.type === 'OptionalCallExpression';
}

/**
 * Pull a slug out of a call: prefer a string-literal argument that names the
 * intent (`setDateType("today")` -> `today`), else the callee name
 * (`onDotButtonClick(i)` -> `dot-button-click`, `router.push()` -> `router`).
 */
function fromCall(call: Extract<Node, { type: 'CallExpression' | 'OptionalCallExpression' }>): string | null {
  for (const arg of call.arguments) {
    if (arg.type === 'StringLiteral' && /[a-zA-Z]/.test(arg.value)) {
      const s = meaningful(slugify(arg.value));
      if (s) return s;
    }
  }
  const callee = call.callee;
  if (callee.type === 'Identifier') {
    return meaningful(slugify(stripHandlerPrefix(callee.name)));
  }
  if (callee.type === 'MemberExpression') {
    if (callee.object.type === 'Identifier') {
      const s = meaningful(slugify(callee.object.name));
      if (s) return s;
    }
    if (callee.property.type === 'Identifier') {
      return meaningful(slugify(stripHandlerPrefix(callee.property.name)));
    }
  }
  return null;
}

/** Depth-first walk returning the slug of the first meaningful call in source order. */
function firstSignalInBody(root: Node): string | null {
  let result: string | null = null;
  const visit = (n: unknown): void => {
    if (result || !n || typeof n !== 'object') return;
    const node = n as Node;
    if (typeof node.type !== 'string') return;
    if (isCall(node)) {
      const s = fromCall(node);
      if (s) {
        result = s;
        return;
      }
    }
    for (const key of Object.keys(node)) {
      if (key === 'loc' || key === 'start' || key === 'end') continue;
      if (key === 'leadingComments' || key === 'trailingComments') continue;
      const child = (node as unknown as Record<string, unknown>)[key];
      if (Array.isArray(child)) {
        for (const c of child) {
          visit(c);
          if (result) return;
        }
      } else {
        visit(child);
      }
      if (result) return;
    }
  };
  visit(root);
  return result;
}

/** Extract a slug from a single handler expression. */
function fromExpression(expr: Expression): string | null {
  switch (expr.type) {
    case 'Identifier':
      return meaningful(slugify(stripHandlerPrefix(expr.name)));
    case 'MemberExpression': {
      // `datePicker.open` -> the receiver names the intent -> `date-picker`.
      if (expr.object.type === 'Identifier') {
        const s = meaningful(slugify(expr.object.name));
        if (s) return s;
      }
      if (expr.property.type === 'Identifier') {
        return meaningful(slugify(stripHandlerPrefix(expr.property.name)));
      }
      return null;
    }
    case 'ArrowFunctionExpression':
    case 'FunctionExpression':
      return firstSignalInBody(expr.body);
    default:
      return null;
  }
}

export function deriveHandlerSignal(open: JSXOpeningElement): string | null {
  for (const key of HANDLER_ATTRS) {
    for (const a of open.attributes) {
      if (a.type !== 'JSXAttribute') continue;
      if (a.name.type !== 'JSXIdentifier' || a.name.name !== key) continue;
      const v = a.value;
      if (!v || v.type !== 'JSXExpressionContainer') continue;
      const expr = v.expression;
      if (expr.type === 'JSXEmptyExpression') continue;
      const s = fromExpression(expr);
      if (s) return s;
    }
  }
  return null;
}

File: packages/swc-plugin-16/src/handler_signal.rs
//! Port of `src/id/handler-signal.ts`, adapted to the SWC AST.
//!
//! Mine an English intent slug from an element's event handlers, so non-Latin
//! UIs (CJK) still get a stable, meaningful id from the handler names/args that
//! developers already write in English.

use swc_core::ecma::ast::{
    ArrowExpr, CallExpr, Callee, Expr, ExprOrSpread, FnExpr, JSXAttrName, JSXAttrOrSpread,
    JSXAttrValue, JSXExpr, JSXOpeningElement, Lit, MemberProp, OptCall,
};
use swc_core::ecma::visit::{Visit, VisitWith};

/// Event handler props, in priority order.
const HANDLER_ATTRS: &[&str] = &["onClick", "onPress", "onChange", "onSubmit", "onValueChange"];

/// Slugs too generic to make a useful id.
const STOPWORDS: &[&str] = &[
    "fn", "cb", "callback", "handler", "handle", "on", "e", "ev", "evt", "event",
];

/// Strip the conventional `handle`/`on` prefix: `handleSubmit` -> `Submit`.
fn strip_handler_prefix(name: &str) -> String {
    let after = name
        .strip_prefix("handle")
        .or_else(|| name.strip_prefix("on"))
        .filter(|rest| {
            rest.chars()
                .next()
                .map_or(false, |c| c.is_ascii_uppercase() || c == '_')
        });
    let stripped = match after {
        Some(rest) => rest.trim_start_matches('_').to_string(),
        None => name.to_string(),
    };
    if stripped.is_empty() {
        name.to_string()
    } else {
        stripped
    }
}

/// Reject empty, single-letter, or stopword slugs.
fn meaningful(slug: String) -> Option<String> {
    if slug.chars().count() < 2 {
        return None;
    }
    let compact = slug.replace('-', "");
    if STOPWORDS.contains(&compact.as_str()) {
        return None;
    }
    Some(slug)
}

/// Slug from a call's args (string literal with a letter) or its callee name.
fn signal_from_call(args: &[ExprOrSpread], callee: Option<&Expr>) -> Option<String> {
    for arg in args {
        if arg.spread.is_some() {
            continue;
        }
        if let Expr::Lit(Lit::Str(s)) = &*arg.expr {
            let sv = s.value.to_string_lossy();
            if sv.chars().any(|c| c.is_ascii_alphabetic()) {
                if let Some(m) = meaningful(crate::slugify::slugify(&sv)) {
                    return Some(m);
                }
            }
        }
    }
    match callee? {
        Expr::Ident(id) => meaningful(crate::slugify::slugify(&strip_handler_prefix(&id.sym))),
        Expr::Member(m) => {
            if let Expr::Ident(obj) = &*m.obj {
                if let Some(x) = meaningful(crate::slugify::slugify(&obj.sym)) {
                    return Some(x);
                }
            }
            if let MemberProp::Ident(prop) = &m.prop {
                return meaningful(crate::slugify::slugify(&strip_handler_prefix(&prop.sym)));
            }
            None
        }
        _ => None,
    }
}

fn from_call_expr(call: &CallExpr) -> Option<String> {
    let callee = match &call.callee {
        Callee::Expr(e) => Some(&**e),
        _ => None,
    };
    signal_from_call(&call.args, callee)
}

fn from_opt_call(call: &OptCall) -> Option<String> {
    signal_from_call(&call.args, Some(&*call.callee))
}

/// Depth-first (source order) finder for the first meaningful call in a body.
struct CallFinder {
    result: Option<String>,
}

impl Visit for CallFinder {
    fn visit_call_expr(&mut self, n: &CallExpr) {
        if self.result.is_some() {
            return;
        }
        if let Some(s) = from_call_expr(n) {
            self.result = Some(s);
            return;
        }
        n.visit_children_with(self);
    }

    fn visit_opt_call(&mut self, n: &OptCall) {
        if self.result.is_some() {
            return;
        }
        if let Some(s) = from_opt_call(n) {
            self.result = Some(s);
            return;
        }
        n.visit_children_with(self);
    }
}

fn first_signal_in_arrow(arrow: &ArrowExpr) -> Option<String> {
    let mut f = CallFinder { result: None };
    arrow.body.visit_with(&mut f);
    f.result
}

fn first_signal_in_fn(func: &FnExpr) -> Option<String> {
    let mut f = CallFinder { result: None };
    if let Some(body) = &func.function.body {
        body.visit_with(&mut f);
    }
    f.result
}

fn from_expression(expr: &Expr) -> Option<String> {
    match expr {
        Expr::Ident(id) => {
            meaningful(crate::slugify::slugify(&strip_handler_prefix(&id.sym)))
        }
        Expr::Member(m) => {
            if let Expr::Ident(obj) = &*m.obj {
                if let Some(x) = meaningful(crate::slugify::slugify(&obj.sym)) {
                    return Some(x);
                }
            }
            if let MemberProp::Ident(prop) = &m.prop {
                return meaningful(crate::slugify::slugify(&strip_handler_prefix(&prop.sym)));
            }
            None
        }
        Expr::Arrow(arrow) => first_signal_in_arrow(arrow),
        Expr::Fn(func) => first_signal_in_fn(func),
        _ => None,
    }
}

pub fn derive_handler_signal(open: &JSXOpeningElement) -> Option<String> {
    for key in HANDLER_ATTRS {
        for attr in &open.attrs {
            let JSXAttrOrSpread::JSXAttr(a) = attr else {
                continue;
            };
            let JSXAttrName::Ident(name) = &a.name else {
                continue;
            };
            if name.sym.as_ref() != *key {
                continue;
            }
            if let Some(JSXAttrValue::JSXExprContainer(c)) = &a.value {
                if let JSXExpr::Expr(expr) = &c.expr {
                    if let Some(s) = from_expression(expr) {
                        return Some(s);
                    }
                }
            }
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn strip_prefix_cases() {
        assert_eq!(strip_handler_prefix("handleSubmit"), "Submit");
        assert_eq!(strip_handler_prefix("onClose"), "Close");
        assert_eq!(strip_handler_prefix("onlyText"), "onlyText"); // lookahead guard
        assert_eq!(strip_handler_prefix("on"), "on");
        assert_eq!(strip_handler_prefix("scrollNext"), "scrollNext");
    }

    #[test]
    fn meaningful_filters() {
        assert_eq!(meaningful("s".to_string()), None);
        assert_eq!(meaningful("fn".to_string()), None);
        assert_eq!(meaningful("today".to_string()), Some("today".to_string()));
    }
}
